Script View

allieglotfelty opened the pull request:
Includes:
- Refactor of script files
- Script View Component
- Script View Container
- Script View Actions 
- Script View Reducers
- Script View tests of all above

allieglotfelty committed:
Add action, component, reducer files for admin to view individual scripts

allieglotfelty committed:
Added component, container, actions, reducer for individual script view with script details and questions.

allieglotfelty committed:
Deleted script view from landing page

allieglotfelty committed:
Remove static script_id = 6 for testing

allieglotfelty committed:
Add test file for script component, update name of script container to view_script.

allieglotfelty committed:
Fix merge conflicts in store and index and merge.

allieglotfelty committed:
Update files related to moving script component into script folder.

allieglotfelty committed:
Add tests for script view component

allieglotfelty committed:
Move script.jsx into script component folder

allieglotfelty committed:
Small update to format of questions in props

allieglotfelty committed:
Add type and payload tests for script setScriptInfo and setScriptQuestion actions

allieglotfelty committed:
Import fetchScript, fetchScriptQuestions, and other modules to mock axios requests

allieglotfelty committed:
Add action tests for script view

allieglotfelty committed:
Play around to address key warning - all tests passing

allieglotfelty committed:
Fix key warning on script tests and refactored script component tests

allieglotfelty committed:
Add tests for script view reducer

allieglotfelty committed:
Fixed merge conflicts

allieglotfelty committed:
Remove MemoryRouter

andioneto commented on public/src/actions/script.jsx:
```diff
@@ -0,0 +1,46 @@
+export function fetchScript(id) {
+  return dispatch => axios.get(`/scripts/${id}`)
+    .then((res) => {
+      const scriptData = res.data;
```
Could use deconstruction for this:
const { data: scriptData } = res;
But totally not necessary.

allieglotfelty replied:
will update so we are consistent!

andioneto commented on public/src/actions/script.jsx:
```diff
@@ -0,0 +1,46 @@
+export function fetchScriptQuestions(id) {
+  return dispatch => axios.get(`/scripts/${id}/scriptQuestions/`)
+    .then((res) => {
+      const scriptQuestions = res.data;
```
Same thing here

allieglotfelty replied:
done.

andioneto commented on public/src/components/script/script.jsx:
```diff
@@ -0,0 +1,44 @@
+    this.props.fetchScriptQuestions(script_id);
+  }
+
+  renderQuestions() {
```
This could be a component instead of a method in this component. Perfect use case for a stateless component.

andioneto commented on public/src/index.jsx:
```diff
@@ -84,6 +86,10 @@ const Root = () => {
               }
             />
             <Route
+              path="/admin/scripts/"
```
Could you add in the isLoggedIn check for consistency?

allieglotfelty replied:
definitely!

andioneto commented on public/src/reducers/script.jsx:
```diff
@@ -0,0 +1,38 @@
+export const SET_SCRIPT_INFO = 'SET_SCRIPT_INFO';
+export const SET_SCRIPT_QUESTIONS = 'SET_SCRIPT_QUESTIONS';
+
+export function scriptInfoReducer(state = scriptInfo, action) {
```
After we combine the actions and sort out integration with the script list view I don't think this will be necessary. Looks good though.

allieglotfelty replied:
Ok. No prob! Good for learning :)

andioneto commented on public/src/store.jsx:
```diff
@@ -19,6 +20,8 @@ const appReducer = combineReducers({
   form: formReducer,
   auth: authStatusReducer,
   account_info: accountInfoReducer,
+  script: scriptInfoReducer,
```
Again, this is a duplicate, but we will have to wait for consolidation. Looks good.

andioneto commented on test/client/actions/script.test.js:
```diff
@@ -0,0 +1,132 @@
+  describe('setScriptInfo', () => {
+    describe('type', () => {
+      it('should have the type "SET_SCRIPT_INFO"', () => {
+        expect(setScriptInfo().type).toEqual('SET_SCRIPT_INFO');
```
Could deconstruct type and payload, but not necessary.

allieglotfelty replied:
done.

andioneto commented on test/client/actions/script.test.js:
```diff
@@ -0,0 +1,132 @@
+    beforeEach(() => {
+      mock = new MockAdapter(axios);
+      mock.onGet('/scripts/1').reply(200, newScript);
+      store = mockStore(initialState);
```
After the first describe statement, could use the suggested code from Steven:
let store;
 +  beforeEach(() => {
 +    store = mockStore(initialState);
 +  });
So you don't have to define a new store in each block.

allieglotfelty replied:
done.

allieglotfelty replied:
also moved the afterEach to the top describe to dry things up a bit.

andioneto commented on test/client/actions/script.test.js:
```diff
@@ -0,0 +1,132 @@
+      store = mockStore(initialState);
+    });
+    afterEach(() => {
+      mock.restore();
```
I meant to ask you, but I ended up switching to .reset bc .restore was causing problems in subsequent tests.

allieglotfelty replied:
ok! reset seems to work fine for mine. Will run with that.

andipants12 commented on test/client/components/script/script.test.js:
```diff
@@ -0,0 +1,136 @@
+  describe('props', () => {
+    describe('script_id prop', () => {
+      it('should be defined', () => {
+        expect(props.script_id).toBeDefined();
```
You could use deconstruction under the first describe block (line 76), so you're not having to reference props in each expect.

allieglotfelty replied:
done!

andioneto commented on test/client/components/script/script.test.js:
```diff
@@ -0,0 +1,136 @@
+        expect(typeof props.script_info).toBe('object');
+      });
+      it('should have 3 properties - name, body, description', () => {
+        const scriptInfoKeys = Object.keys(props.script_info);
```
I wrote a helper function for this if you want to take a look. Uses reduce and checks if the object has specific properties.

allieglotfelty replied:
Yea I saw that in your PR - will work it in here.

andioneto reviewed (commented):
Let me know what you think of comments, I'll look for responses and approve 😸

allieglotfelty committed:
Deconstruct res.data

allieglotfelty committed:
Add isLogin check for scriptpage

allieglotfelty committed:
Update property tests for script component and actions

allieglotfelty committed:
Correct path for admin/scripts/:id and update deconstruction in script action

allieglotfelty committed:
Refactor script view to integrate with script_list view and add functional component for rendering script questions

allieglotfelty committed:
Rename script_question_list component

allieglotfelty committed:
Delete console.log statement

allieglotfelty committed:
Update of script view tests, Remove scriptInfo reducer from reducers, actions, store, Update script view to render with current_script

allieglotfelty committed:
Update merge conflicts and merge

allieglotfelty committed:
Consolidate destructuring of questions and current_script

allieglotfelty committed:
Delete changes made for testing script view

allieglotfelty committed:
Added login check back to script view route

allieglotfelty committed:
Delete .purge from persistStore

allieglotfelty committed:
Add export of defaultScriptQuestions

allieglotfelty committed:
Refactor with new client_test_helpers

allieglotfelty committed:
Added fixtures for script view tests

allieglotfelty committed:
Added fixtures for script view tests

allieglotfelty committed:
Added fixtures for script view tests

allieglotfelty committed:
Remove space

allieglotfelty committed:
Update script_question_list with client helpers

allieglotfelty committed:
Merge branch 'master' into agScriptView

JadziaHax reviewed (dismissed):
👍

andioneto commented on test/client/client_fixtures.js:
```diff
@@ -206,6 +206,66 @@ export default {
     }
+  },
+  scriptViewFixtures: {
+    defaultScript: {
```
Might be nice to import the default script obj from the source file (reducer), rather than hard code this here.

allieglotfelty replied:
I actually removed the set_script reducer, so I'll just remove this all together.

andioneto commented on test/client/components/script/script.test.js:
```diff
@@ -0,0 +1,79 @@
+
+const props = {
+  id: 1,
+  scripts,
```
Why is scripts here? I don't see where it's being tested and where it's used in the component...

allieglotfelty replied:
I was using it to access the current_script the way we would through admin_scripts in the regular code, but I think you're right that I should just be testing current_script.

andioneto commented on test/client/components/script/script_question_list.test.js:
```diff
@@ -0,0 +1,30 @@
+};
+
+describe('QuestionList component', () => {
+  const { questions } = props;
```
I think when checking props in component test it might be better to shallowly render the component. Otherwise you're just confirming that the constant "props" you defined it defined.

allieglotfelty committed:
Remove defaultScript

allieglotfelty committed:
Update admin_scripts to current_script

allieglotfelty committed:
Add component shallow render to test props properly

allieglotfelty committed:
Merge with master.

Merge branch 'master' of https://github.com/Twilio-org/phonebank into agScriptView

allieglotfelty committed:
merge with PR changes.
Merge branch 'agScriptView' of https://github.com/Twilio-org/phonebank into agScriptView

allieglotfelty committed:
Clarify two test descriptions

allieglotfelty reviewed (commented):
Thanks for the comments, @andipants12! Can you take a look at my fixes and approve/comment more based on what you think?

andioneto commented on test/client/components/script/script.test.js:
```diff
@@ -0,0 +1,78 @@
+  fetchScriptQuestions
+};
+
+describe('<ScriptPage />', () => {
```
Are you going to add in a shallow render and then test the props or skip that for now?
@allieglotfelty

allieglotfelty committed:
Add component rendering tests to ScriptPage component

andioneto reviewed (approved):
Looks goodie to me.
